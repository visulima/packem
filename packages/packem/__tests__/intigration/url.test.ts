import type { Dirent } from "node:fs";
import { cpSync } from "node:fs";
import { readdir, rm } from "node:fs/promises";

import { ensureDir, isAccessibleSync, readFileSync, writeFile, writeJson } from "@visulima/fs";
import { basename, join, resolve } from "@visulima/path";
import { temporaryDirectory } from "tempy";
import { afterEach, beforeEach, describe, expect, it } from "vitest";

import type { UrlOptions } from "../../src/rollup/plugins/url";
import { createPackageJson, createPackemConfig, execPackem } from "../helpers";

const fixturePath = join(__dirname, "../..", "__fixtures__", "url");

describe("url", () => {
    let temporaryDirectoryPath: string;

    beforeEach(async () => {
        temporaryDirectoryPath = temporaryDirectory();
    });

    afterEach(async () => {
        await rm(temporaryDirectoryPath, { recursive: true });
    });

    const build = async (type: string, options: Partial<UrlOptions>, generatedFiles: string[], useSnapshot?: boolean) => {
        // copy fixtures to temporary directory
        cpSync(join(fixturePath, `${type}.js`), join(temporaryDirectoryPath, "src", `${type}.js`));
        cpSync(join(fixturePath, `${type}.${type}`), join(temporaryDirectoryPath, "src", `${type}.${type}`));

        await createPackageJson(temporaryDirectoryPath, {
            exports: {
                ".": {
                    import: `./dist/${type}.mjs`,
                    require: `./dist/${type}.cjs`,
                },
            },
        });

        await createPackemConfig(temporaryDirectoryPath, {
            config: {
                rollup: {
                    url: options,
                },
            },
            transformer: "esbuild",
        });

        const binProcess = await execPackem("build", [], {
            cwd: temporaryDirectoryPath,
            reject: false,
        });

        expect(binProcess.stderr).toBe("");
        expect(binProcess.exitCode).toBe(0);

        if (useSnapshot !== false) {
            const mjsContent = readFileSync(join(temporaryDirectoryPath, "dist", `${type}.mjs`));

            expect(mjsContent).toMatchSnapshot("mjs");

            const cjsContent = readFileSync(join(temporaryDirectoryPath, "dist", `${type}.cjs`));

            expect(cjsContent).toMatchSnapshot("cjs");
        }

        const distributionPath = join(temporaryDirectoryPath, "dist");

        const foundFiles: Dirent[] = await readdir(distributionPath, {
            recursive: true,
            withFileTypes: true,
        });

        const files: string[] = foundFiles
            .filter((dirent) => dirent.isFile())
            // @TODO: Change this readdir to @visulima/fs readdir
            // eslint-disable-next-line deprecation/deprecation
            .map((dirent) => join(dirent.path, dirent.name));

        expect(files).toStrictEqual(generatedFiles);
    };

    it("should inline png files", async () => {
        expect.assertions(5);

        await build("png", { limit: 10 * 1024 }, [join(temporaryDirectoryPath, "/dist/png.cjs"), join(temporaryDirectoryPath, "/dist/png.mjs")]);
    });

    it("should inline jpg files", async () => {
        expect.assertions(5);

        await build("jpg", { limit: 10 * 1024 }, [join(temporaryDirectoryPath, "/dist/jpg.cjs"), join(temporaryDirectoryPath, "/dist/jpg.mjs")]);
    });

    it("should inline jpeg files", async () => {
        expect.assertions(5);

        await build("jpeg", { limit: 10 * 1024 }, [join(temporaryDirectoryPath, "/dist/jpeg.cjs"), join(temporaryDirectoryPath, "/dist/jpeg.mjs")]);
    });

    it("should inline gif files", async () => {
        expect.assertions(5);

        await build("gif", { limit: 10 * 1024 }, [join(temporaryDirectoryPath, "/dist/gif.cjs"), join(temporaryDirectoryPath, "/dist/gif.mjs")]);
    });

    it("should inline webp files", async () => {
        expect.assertions(5);

        await build("webp", { limit: 10 * 1024 }, [join(temporaryDirectoryPath, "/dist/webp.cjs"), join(temporaryDirectoryPath, "/dist/webp.mjs")]);
    });

    it("should inline text files", async () => {
        expect.assertions(5);

        await build("svg", { limit: 10 * 1024 }, [join(temporaryDirectoryPath, "/dist/svg.cjs"), join(temporaryDirectoryPath, "/dist/svg.mjs")]);
    });

    it("inline \"large\" files", async () => {
        expect.assertions(5);

        await build("svg", { limit: 10 }, [
            join(temporaryDirectoryPath, "/dist/98ea1a8cc8cd9baf.svg"),
            join(temporaryDirectoryPath, "/dist/svg.cjs"),
            join(temporaryDirectoryPath, "/dist/svg.mjs"),
        ]);
    });

    it("should limit: 0, emitFiles: false, publicPath: empty", async () => {
        expect.assertions(5);

        await build("svg", { emitFiles: false, limit: 0, publicPath: "" }, [
            join(temporaryDirectoryPath, "/dist/svg.cjs"),
            join(temporaryDirectoryPath, "/dist/svg.mjs"),
        ]);
    });

    it("should copy files, limit: 0", async () => {
        expect.assertions(5);

        await build("svg", { emitFiles: true, limit: 0 }, [
            join(temporaryDirectoryPath, "/dist/98ea1a8cc8cd9baf.svg"),
            join(temporaryDirectoryPath, "/dist/svg.cjs"),
            join(temporaryDirectoryPath, "/dist/svg.mjs"),
        ]);
    });

    it("copy \"large\" binary files, limit: 10", async () => {
        expect.assertions(5);

        await build("svg", { emitFiles: true, limit: 10 }, [
            join(temporaryDirectoryPath, "/dist/98ea1a8cc8cd9baf.svg"),
            join(temporaryDirectoryPath, "/dist/svg.cjs"),
            join(temporaryDirectoryPath, "/dist/svg.mjs"),
        ]);
    });

    it("should copy files with include by absolute path, limit: 0", async () => {
        expect.assertions(5);

        await build(
            "svg",
            {
                emitFiles: true,
                include: [resolve(fixturePath, "*.svg")],
                limit: 0,
            },
            [
                join(temporaryDirectoryPath, "/dist/98ea1a8cc8cd9baf.svg"),
                join(temporaryDirectoryPath, "/dist/svg.cjs"),
                join(temporaryDirectoryPath, "/dist/svg.mjs"),
            ],
        );
    });

    it("should use publicPath", async () => {
        expect.assertions(5);

        await build("png", { limit: 10, publicPath: "/batman/" }, [
            join(temporaryDirectoryPath, "/dist/6b71fbe07b498a82.png"),
            join(temporaryDirectoryPath, "/dist/png.cjs"),
            join(temporaryDirectoryPath, "/dist/png.mjs"),
        ]);
    });

    it("should use publicPath with file that has empty dirname", async () => {
        expect.assertions(5);

        const type = "png";

        await build(
            type,
            {
                emitFiles: true,
                fileName: "[dirname][hash][extname]",
                limit: 10,
                publicPath: "/batman/",
                sourceDir: join(temporaryDirectoryPath, "src"),
            },
            [
                join(temporaryDirectoryPath, "/dist/6b71fbe07b498a82.png"),
                join(temporaryDirectoryPath, "/dist/png.cjs"),
                join(temporaryDirectoryPath, "/dist/png.mjs"),
            ],
            false,
        );

        const mjsContent = readFileSync(join(temporaryDirectoryPath, "dist", `${type}.mjs`));

        expect(mjsContent).toBe(`const png = "/batman/6b71fbe07b498a82.png";

export { png as default };
`);

        const cjsContent = readFileSync(join(temporaryDirectoryPath, "dist", `${type}.cjs`));

        expect(cjsContent).toBe(`'use strict';

const png = "/batman/6b71fbe07b498a82.png";

module.exports = png;
`);
    });

    it("should create a nested directory for the output, if required", async () => {
        expect.assertions(5);

        await build("png", { emitFiles: true, fileName: "joker/[hash][extname]", limit: 10 }, [
            join(temporaryDirectoryPath, "/dist/png.cjs"),
            join(temporaryDirectoryPath, "/dist/png.mjs"),
            join(temporaryDirectoryPath, "/dist/joker/6b71fbe07b498a82.png"),
        ]);
    });

    it("should create a file with the name and extension of the file", async () => {
        expect.assertions(5);

        await build("png", { emitFiles: true, fileName: "[name][extname]", limit: 10 }, [
            join(temporaryDirectoryPath, "/dist/png.cjs"),
            join(temporaryDirectoryPath, "/dist/png.mjs"),
            join(temporaryDirectoryPath, "/dist/png.png"),
        ]);
    });

    it("should create a file with the name, hash and extension of the file", async () => {
        expect.assertions(5);

        await build("png", { emitFiles: true, fileName: "[name]-[hash][extname]", limit: 10 }, [
            join(temporaryDirectoryPath, "/dist/png-6b71fbe07b498a82.png"),
            join(temporaryDirectoryPath, "/dist/png.cjs"),
            join(temporaryDirectoryPath, "/dist/png.mjs"),
        ]);
    });

    it("should prefix the file with the parent directory of the source file", async () => {
        expect.assertions(5);

        await build("png", { emitFiles: true, fileName: "[dirname][hash][extname]", limit: 10 }, [
            join(temporaryDirectoryPath, "/dist/png.cjs"),
            join(temporaryDirectoryPath, "/dist/png.mjs"),
            join(temporaryDirectoryPath, "/dist/src/6b71fbe07b498a82.png"),
        ]);
    });

    it("should prefix the file with the parent directory of the source file, relative to the sourceDir option", async () => {
        expect.assertions(5);

        const type = "png";

        const pngPath = join(basename(temporaryDirectoryPath), "/src/6b71fbe07b498a82.png");

        await build(
            type,
            {
                emitFiles: true,
                fileName: "[dirname][hash][extname]",
                limit: 10,
                sourceDir: join(temporaryDirectoryPath, ".."),
            },
            [join(temporaryDirectoryPath, "/dist/png.cjs"), join(temporaryDirectoryPath, "/dist/png.mjs"), join(temporaryDirectoryPath, "/dist/", pngPath)],
            false,
        );

        const mjsContent = readFileSync(join(temporaryDirectoryPath, "dist", `${type}.mjs`));

        expect(mjsContent).toBe(`const png = "${pngPath}";

export { png as default };
`);

        const cjsContent = readFileSync(join(temporaryDirectoryPath, "dist", `${type}.cjs`));

        expect(cjsContent).toBe(`'use strict';

const png = "${pngPath}";

module.exports = png;
`);
    });

    it("should copy the file according to destDir option", async () => {
        expect.assertions(6);

        await build(
            "png",
            {
                destDir: join(temporaryDirectoryPath, "output/dest"),
                emitFiles: true,
                fileName: "[dirname][hash][extname]",
                limit: 10,
            },
            [join(temporaryDirectoryPath, "/dist/png.cjs"), join(temporaryDirectoryPath, "/dist/png.mjs")],
        );

        expect(isAccessibleSync(join(temporaryDirectoryPath, "output/dest/src/6b71fbe07b498a82.png"))).toBe(true);
    });

    it("should import images from the node_modules", async () => {
        const type = "svg";

        await writeFile(
            join(temporaryDirectoryPath, "src", `${type}.js`),
            `import svg from "@test/images/icons/${type}.${type}";

export default svg;`,
        );

        const imagesPackagePath = join(temporaryDirectoryPath, "node_modules/@test/images");

        await ensureDir(imagesPackagePath);

        cpSync(join(fixturePath, `${type}.${type}`), join(imagesPackagePath, "icons", `${type}.${type}`));

        await writeJson(
            join(imagesPackagePath, "package.json"),
            {
                name: "@test/images",
            },
            {
                recursive: true,
            },
        );

        await createPackageJson(temporaryDirectoryPath, {
            exports: {
                ".": {
                    import: `./dist/${type}.mjs`,
                    require: `./dist/${type}.cjs`,
                },
            },
        });

        await createPackemConfig(temporaryDirectoryPath);

        const binProcess = await execPackem("build", [], {
            cwd: temporaryDirectoryPath,
            reject: false,
        });

        expect(binProcess.stderr).toBe("");
        expect(binProcess.exitCode).toBe(0);

        const mjsContent = readFileSync(join(temporaryDirectoryPath, "dist", `${type}.mjs`));

        expect(mjsContent).toMatchSnapshot("mjs");

        const cjsContent = readFileSync(join(temporaryDirectoryPath, "dist", `${type}.cjs`));

        expect(cjsContent).toMatchSnapshot("cjs");
    });
});
