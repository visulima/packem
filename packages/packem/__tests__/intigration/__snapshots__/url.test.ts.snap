// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`url > copy "large" binary files, limit: 10 > cjs 1`] = `
"'use strict';

const svg = "98ea1a8cc8cd9baf.svg";

module.exports = svg;
"
`;

exports[`url > copy "large" binary files, limit: 10 > mjs 1`] = `
"const svg = "98ea1a8cc8cd9baf.svg";

export { svg as default };
"
`;

exports[`url > inline "large" files > cjs 1`] = `
"'use strict';

const svg = "98ea1a8cc8cd9baf.svg";

module.exports = svg;
"
`;

exports[`url > inline "large" files > mjs 1`] = `
"const svg = "98ea1a8cc8cd9baf.svg";

export { svg as default };
"
`;

exports[`url > should copy files with include by absolute path, limit: 0 > cjs 1`] = `
"'use strict';

const svg = "98ea1a8cc8cd9baf.svg";

module.exports = svg;
"
`;

exports[`url > should copy files with include by absolute path, limit: 0 > mjs 1`] = `
"const svg = "98ea1a8cc8cd9baf.svg";

export { svg as default };
"
`;

exports[`url > should copy files, limit: 0 > cjs 1`] = `
"'use strict';

const svg = "98ea1a8cc8cd9baf.svg";

module.exports = svg;
"
`;

exports[`url > should copy files, limit: 0 > mjs 1`] = `
"const svg = "98ea1a8cc8cd9baf.svg";

export { svg as default };
"
`;

exports[`url > should copy the file according to destDir option > cjs 1`] = `
"'use strict';

const png = "src/6b71fbe07b498a82.png";

module.exports = png;
"
`;

exports[`url > should copy the file according to destDir option > mjs 1`] = `
"const png = "src/6b71fbe07b498a82.png";

export { png as default };
"
`;

exports[`url > should create a file with the name and extension of the file > cjs 1`] = `
"'use strict';

const png = "png.png";

module.exports = png;
"
`;

exports[`url > should create a file with the name and extension of the file > mjs 1`] = `
"const png = "png.png";

export { png as default };
"
`;

exports[`url > should create a file with the name, hash and extension of the file > cjs 1`] = `
"'use strict';

const png = "png-6b71fbe07b498a82.png";

module.exports = png;
"
`;

exports[`url > should create a file with the name, hash and extension of the file > mjs 1`] = `
"const png = "png-6b71fbe07b498a82.png";

export { png as default };
"
`;

exports[`url > should create a nested directory for the output, if required > cjs 1`] = `
"'use strict';

const png = "joker/6b71fbe07b498a82.png";

module.exports = png;
"
`;

exports[`url > should create a nested directory for the output, if required > mjs 1`] = `
"const png = "joker/6b71fbe07b498a82.png";

export { png as default };
"
`;

exports[`url > should import images from the node_modules > cjs 1`] = `
"'use strict';

const svg = "data:image/svg+xml;base64,PHN2Zz48cGF0aCBkPSIiLz48L3N2Zz4=";

module.exports = svg;
"
`;

exports[`url > should import images from the node_modules > mjs 1`] = `
"const svg = "data:image/svg+xml;base64,PHN2Zz48cGF0aCBkPSIiLz48L3N2Zz4=";

export { svg as default };
"
`;

exports[`url > should inline gif files > cjs 1`] = `
"'use strict';

const gif = "data:image/gif;base64,R0lGODlhAQABAPAAAP///wAAACH5BAAAAAAAIf8LSW1hZ2VNYWdpY2sOZ2FtbWE9MC40NTQ1NDUALAAAAAABAAEAAAICRAEAOw==";

module.exports = gif;
"
`;

exports[`url > should inline gif files > mjs 1`] = `
"const gif = "data:image/gif;base64,R0lGODlhAQABAPAAAP///wAAACH5BAAAAAAAIf8LSW1hZ2VNYWdpY2sOZ2FtbWE9MC40NTQ1NDUALAAAAAABAAEAAAICRAEAOw==";

export { gif as default };
"
`;

exports[`url > should inline jpeg files > cjs 1`] = `
"'use strict';

const jpeg = "data:image/jpeg;base64,/9j/4AAQSkZJRgABAQAAAQABAAD/2wBDAFA3PEY8MlBGQUZaVVBfeMiCeG5uePWvuZHI////////////////////////////////////////////////////wAALCAABAAEBAREA/8QAFAABAAAAAAAAAAAAAAAAAAAABP/EABQQAQAAAAAAAAAAAAAAAAAAAAD/2gAIAQEAAD8Aa//Z";

module.exports = jpeg;
"
`;

exports[`url > should inline jpeg files > mjs 1`] = `
"const jpeg = "data:image/jpeg;base64,/9j/4AAQSkZJRgABAQAAAQABAAD/2wBDAFA3PEY8MlBGQUZaVVBfeMiCeG5uePWvuZHI////////////////////////////////////////////////////wAALCAABAAEBAREA/8QAFAABAAAAAAAAAAAAAAAAAAAABP/EABQQAQAAAAAAAAAAAAAAAAAAAAD/2gAIAQEAAD8Aa//Z";

export { jpeg as default };
"
`;

exports[`url > should inline jpg files > cjs 1`] = `
"'use strict';

const jpg = "data:image/jpeg;base64,/9j/4AAQSkZJRgABAQAAAQABAAD/2wBDAFA3PEY8MlBGQUZaVVBfeMiCeG5uePWvuZHI////////////////////////////////////////////////////wAALCAABAAEBAREA/8QAFAABAAAAAAAAAAAAAAAAAAAABP/EABQQAQAAAAAAAAAAAAAAAAAAAAD/2gAIAQEAAD8Aa//Z";

module.exports = jpg;
"
`;

exports[`url > should inline jpg files > mjs 1`] = `
"const jpg = "data:image/jpeg;base64,/9j/4AAQSkZJRgABAQAAAQABAAD/2wBDAFA3PEY8MlBGQUZaVVBfeMiCeG5uePWvuZHI////////////////////////////////////////////////////wAALCAABAAEBAREA/8QAFAABAAAAAAAAAAAAAAAAAAAABP/EABQQAQAAAAAAAAAAAAAAAAAAAAD/2gAIAQEAAD8Aa//Z";

export { jpg as default };
"
`;

exports[`url > should inline png files > cjs 1`] = `
"'use strict';

const png = "data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAAAAAA6fptVAAAACklEQVR4nGP6DwABBQECz6AuzQAAAABJRU5ErkJggg==";

module.exports = png;
"
`;

exports[`url > should inline png files > mjs 1`] = `
"const png = "data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAAAAAA6fptVAAAACklEQVR4nGP6DwABBQECz6AuzQAAAABJRU5ErkJggg==";

export { png as default };
"
`;

exports[`url > should inline text files > cjs 1`] = `
"'use strict';

const svg = "data:image/svg+xml;base64,PHN2Zz48cGF0aCBkPSIiLz48L3N2Zz4=";

module.exports = svg;
"
`;

exports[`url > should inline text files > mjs 1`] = `
"const svg = "data:image/svg+xml;base64,PHN2Zz48cGF0aCBkPSIiLz48L3N2Zz4=";

export { svg as default };
"
`;

exports[`url > should inline webp files > cjs 1`] = `
"'use strict';

const webp = "data:image/webp;base64,UklGRiQAAABXRUJQVlA4IBgAAAAwAQCdASoBAAEAB0CWJaQAA3AA/u9gAAA=";

module.exports = webp;
"
`;

exports[`url > should inline webp files > mjs 1`] = `
"const webp = "data:image/webp;base64,UklGRiQAAABXRUJQVlA4IBgAAAAwAQCdASoBAAEAB0CWJaQAA3AA/u9gAAA=";

export { webp as default };
"
`;

exports[`url > should limit: 0, emitFiles: false, publicPath: empty > cjs 1`] = `
"'use strict';

const svg = "98ea1a8cc8cd9baf.svg";

module.exports = svg;
"
`;

exports[`url > should limit: 0, emitFiles: false, publicPath: empty > mjs 1`] = `
"const svg = "98ea1a8cc8cd9baf.svg";

export { svg as default };
"
`;

exports[`url > should prefix the file with the parent directory of the source file > cjs 1`] = `
"'use strict';

const png = "src/6b71fbe07b498a82.png";

module.exports = png;
"
`;

exports[`url > should prefix the file with the parent directory of the source file > mjs 1`] = `
"const png = "src/6b71fbe07b498a82.png";

export { png as default };
"
`;

exports[`url > should use publicPath > cjs 1`] = `
"'use strict';

const png = "/batman/6b71fbe07b498a82.png";

module.exports = png;
"
`;

exports[`url > should use publicPath > mjs 1`] = `
"const png = "/batman/6b71fbe07b498a82.png";

export { png as default };
"
`;
